added functions to get freq/phase/amp/att for ad99xx and urukul

lriesebos opened the pull request:
Signed-off-by: Leon Riesebos <leon.riesebos@duke.edu>

<!--

Thank you for submitting a PR to ARTIQ!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

You can also read more about contributing to ARTIQ in this document:
https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#contributing-code

Based on https://raw.githubusercontent.com/PyCQA/pylint/master/.github/PULL_REQUEST_TEMPLATE.md
-->

# ARTIQ Pull Request

## Description of Changes

Added `get()` and `get_mu()`, `get_att()`, and `get_att_mu()` functions for the AD9910 and AD9912 coredevice drivers.

### Related Issue

<!-- 
If this PR fixes a particular issue, use the following to automatically close that issue
once this PR gets merged:

Closes #XXX 
-->

Closes #1616 

## Type of Changes

<!-- Leave ONLY the corresponding lines for the applicable type of change: -->
|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

## Steps (Choose relevant, delete irrelevant before submitting)

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Update [RELEASE_NOTES.rst](../RELEASE_NOTES.rst) if there are noteworthy changes, especially if there are changes to existing APIs.
- [x] Close/update issues.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments
- [x] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Documentation Changes

- [x] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`cd doc/manual/; make html`) to ensure no errors.

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

lriesebos commented:
@fanmingyu212 , maybe you can test these changes?

fanmingyu212 reviewed (commented):
See https://gist.github.com/fanmingyu212/8b5de74d45334c0338bc30223b221a26 for the test code. In short, all functions in ad9910.py and urukul.py are tested. ad9912.py are not tested as I don't have the hardware. The gateware version used for testing is 7.7581.e577542f.beta.

`get()`, `get_mu()`, `get_att_mu()`, and `get_att()` in ad9910.py are tested.

`set_att()`, `get_att_mu_()`, and `get_att()` in urukul.py are also tested.

`mu_to_att()` and `att_to_mu()` in urukul.py are confirmed to be inverse functions of each other, but I don't have a great way to explicitly test the result here.

airwoodix commented:
@lriesebos Thanks a lot for the patch. I think the following points should be addressed:

- type hints: I guess that they're not strictly needed, or are they? I see their relevance, especially for `get()` and think they make a lot of sense in the coredevice area. I'd however add them everywhere or not at all;
- It'd be great to also add counterparts to `set_frequency()` & co for reading the FTW, POW and ASF registers (@0x7-0x9);
- would also be nice to add @fanmingyu212 's tests in `test/coredevice/test_ad9910.py`;
- what's the rationale behind `cpld.get_att_mu_()`? Are there situations where you can't afford the overhead of cycling once through the shift register? The SR management is already confusing for newcomers (see forum), does this help or add more confusion? With this patch `get_att_mu()` returns the hardware value but `get_att()` returns the cached value.

lriesebos commented:
@fanmingyu212 thanks for testing! I have merged parts of your testing code into the ARTIQ tests in a way that fits the existing format, see eda2a04 . Though I am not able to run these test cases here, so I guess CI will tell us if there are bugs in the test code.

For the other points:
- @airwoodix I think type hints are not required, but they do give the compiler direct instructions on what types to infer. I found it especially useful for functions that return a `TInt64` as it will not be implicitly casted to a `TInt32` when assigned to a variable. For parameters, not typing them might allow the user to put in different types (e.g. a `TInt32` or a `TInt64). I am both fine with typing everything or with minimal typing. @sbourdeauducq , what do you recommend/prefer?
- I added get functions for the io update registers, see fe7bc94 .
- For `cpld.get_att_mu_()`, my idea was to not add any new complexity to the existing confusion around the shift register. New users already have to learn to call `get_att_mu()` once before doing other attenuation operations and that remains the same. Though if others see this differently, I am happy to change it. I see the following options:
   1. Keep the existing structure, users need to call `get_att_mu()` before they can use `get_att_mu_()`
   2. `get_att_mu_()` will always call `get_att_mu()` first, so now users can call one of both to cache the register value
   3. Add a flag `get_att_mu_(cached=True)`, now users can choose between 1 and 2, for example with 1 as default

sbourdeauducq commented:
>     * For parameters, not typing them might allow the user to put in different types (e.g. a `TInt32` or a `TInt64`). I am both fine with typing everything or with minimal typing. @sbourdeauducq , what do you recommend/prefer?

Put the type annotations on kernels, NAC3 (the new compiler) will require them.

lriesebos commented:
@airwoodix , I implemented your suggestion, `get_att_mu_()` now reads the shift register before returning the value. I also refactored the three files of interest and added typing for all kernel functions.

@fanmingyu212 could you be so kind to test the ad9910 and urukul driver in its completeness after my refactoring and typing?

fanmingyu212 commented on artiq/coredevice/ad9912.py:
```diff
@@ -123,7 +123,7 @@ def set_att_mu(self, att):
 
     @kernel
-    def set_att(self, att):
+    def set_att(self, att: TInt32):
```
Should the type here be TFloat instead of TInt32?

fanmingyu212 commented on artiq/coredevice/urukul.py:
```diff
@@ -343,13 +365,41 @@ def get_att_mu(self):
+        :return: 8-bit digital attenuation setting:
+            255 minimum attenuation, 0 maximum attenuation (31.5 dB)
+        """
+        return int32((self.get_att_mu >> (channel * 8)) & 0xff)
```
Should it be `self.get_att_mu()` instead of `self.get_att_mu` here?

fanmingyu212 reviewed (changes requested):
@lriesebos I did testing with AD9910 and found 2 possible bugs as commented. After changing those two lines, the [test code](https://gist.github.com/fanmingyu212/8b5de74d45334c0338bc30223b221a26) (only slightly modified in the `test_get_att` function with 1 ms delay between `ad9910.get_att` and `ad9910.get_att_mu` to prevent a RTIO underflow) runs with no problem.

lriesebos commented:
@fanmingyu212 thanks again for testing!

@sbourdeauducq do we need to do anything else to get this merged?

airwoodix commented on .gitignore:
```diff
@@ -9,6 +9,7 @@ __pycache__/
 *.pcap
 *.prof
 .ipynb_checkpoints
+/.idea/
```
This should go to your ``.git/info/exclude`` or your global gitignore (``config core.excludesFile``).

airwoodix commented on artiq/coredevice/urukul.py:
```diff
@@ -343,13 +365,41 @@ def get_att_mu(self):
 
     @kernel
-    def set_sync_div(self, div):
+    def get_att_mu_(self, channel: TInt32) -> TInt32:
```
This should have a more explicit name. Maybe ``get_channel_att_mu``?

airwoodix commented on artiq/coredevice/urukul.py:
```diff
@@ -343,13 +365,41 @@ def get_att_mu(self):
+        return int32((self.get_att_mu() >> (channel * 8)) & 0xff)
+
+    @kernel
+    def get_att(self, channel: TInt32) -> TFloat:
```
For consistency, would it make sense to make this return all the attenuations in dB and add ``get_channel_att()``? The symmetry between machine and human units wouldn't be perfect (I guess that return the 32 bits of the SR in ``get_att_mu()`` is more appropriate than a tuple) but IMHO better.

lriesebos replied:
I renamed the function. `get_att()` can not be implemented without storing extra information because a locally generated list can not be returned from a function (i.e. object lifetime). To implement this, there needs to be a second `att_reg` that stores the float values. That sounds a bit overkill to me and I think we can leave that function out. Do you agree?

airwoodix replied:
Thanks! I agree: I guess that most code will use ``get_channel_att`` anyway.

airwoodix commented:
@lriesebos Thanks for the feedback and the changes. The additions are really great! I just think that the attenuator getter interface should be less cryptic. What do you think about changing the function names as suggested?

fanmingyu212 commented:
The latest changes also run with no problem on our setup (ad9910).

sbourdeauducq commented:
Can you clean up the commit list a bit? For example, make it only two commits, the first one with the style changes in the existing code, and the second one that adds the functions.

lriesebos commented:
@sbourdeauducq commits are cleaned up as suggested. if there is anything else you would like me to do, let me know.

sbourdeauducq commented:
Thanks. I assume this is tested.
